src/lib/logical_query_plan/lqp_utils.cpp

```diff
@@ -506,4 +526,32 @@ void remove_invalid_fds(const std::shared_ptr<const AbstractLQPNode>& lqp, std::
    */
 }
 
+std::shared_ptr<AbstractLQPNode> find_diamond_origin_node(
+    const std::shared_ptr<AbstractLQPNode>& union_root_node) {
+  Assert(union_root_node->type == LQPNodeType::Union, "Expecting UnionNode as the diamond's root node.");
+  DebugAssert(union_root_node->input_count() > 1, "Diamond root node does not have two inputs.");
+  bool is_diamond = true;
+  std::optional<std::shared_ptr<AbstractLQPNode>> diamond_origin_node;
+  visit_lqp(union_root_node, [&](const auto& diamond_node) {
+    if (!is_diamond) return LQPVisitation::DoNotVisitInputs;
+    if (diamond_node->output_count() > 1) {
+      if (!diamond_origin_node.has_value()) {
+        diamond_origin_node = diamond_node;
+      } else if (diamond_origin_node != diamond_node) {
```

Comment:
Does this condition ever evaluate to false?  Doc of `visit_lqp` in `lqp_utils.hpp`: _The algorithm is breadth-first search. Each node is visited exactly once._ Thus, `diamond_node` should not be a node we've seen before.